### PR TITLE
[ExportSystemC] Support more emitc operations and types

### PR DESCRIPTION
**Pull request by maerhart**
* Implement emission patterns for all emitc operations except `variable` for which we have our own implementation. Support for template arguments to function calls is not added yet, but can be easily added when there is a need.
* Add emission pattern support for attributes: this is helpful for operations like `emitc.constant` which can take any attribute and print it. Having them pattern based allows for more modularity and reuse.
* Move the helper function to emit parentheses around an expression to the `InlineEmitter` to share it across the various files implementing emission patterns and allow for easier use.

### CODE REVIEW

File: lib/Target/ExportSystemC/Patterns/EmitCEmissionPatterns.cpp [hunk 1]
```diff
@@ -26,11 +27,181 @@ namespace {
...
+    if (value.getDefiningOp<CastOp>())
+      return Precedence::CAST;
+    return {};
+  }
+
+  void emitInlined(Value value, EmissionPrinter &p) override {
+    auto castOp = value.getDefiningOp<CastOp>();
+    p << "(";
+    p.emitType(castOp.getDest().getType());
+    p << ") ";
```
Comment (seldridge) [inline lib/Target/ExportSystemC/Patterns/EmitCEmissionPatterns.cpp:151]: This is a C-style cast. This seems fine for `emitc::CastOp`. Are there situations where this should be emitted as a C++-style `static_cast`?

File: test/Target/ExportSystemC/basic.mlir [hunk 2]
```diff
@@ -133,4 +133,27 @@ systemc.module @systemCTypes (%p0: !systemc.in<!systemc.int_base>,
...
+
+    // CHECK-NEXT: size_t idx;
+    %idx = systemc.cpp.variable : index
+    // CHECK-NEXT: int* somePtr = (int*) malloc(4);
+    %4 = hw.constant 4 : i32
+    %5 = emitc.call "malloc" (%4) : (i32) -> !emitc.ptr<!emitc.opaque<"void">>
+    %6 = emitc.cast %5 : !emitc.ptr<!emitc.opaque<"void">> to !emitc.ptr<!emitc.opaque<"int">>
+    %somePtr = systemc.cpp.variable %6 : !emitc.ptr<!emitc.opaque<"int">>
+  }
+}
```
Comment (seldridge) [inline test/Target/ExportSystemC/basic.mlir:157]: This is a nice, terse example of the features added.

Are there any concerns about maintainability of this test going forward as more `emitc` features are added? I.e., would this test be more maintainable as a series of smaller checks on each feature?

**Review by seldridge — APPROVED** [review]
LGTM

Nitty questions here and there.

File: lib/Target/ExportSystemC/Patterns/BuiltinEmissionPatterns.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      p << "bool";
+      break;
+    case 8:
+    case 16:
+    case 32:
+    case 64:
+      p << (type.isSigned() ? "" : "u") << "int" << bitWidth << "_t";
+      break;
+    default:
+      assert(false && "All cases allowed by match function must be covered.");
```
[the remark this answers is not in the document]
Comment (maerhart) [reply lib/Target/ExportSystemC/Patterns/BuiltinEmissionPatterns.cpp:69]: I can promote this to a proper error message via the diagnostics engine, but when this case it hit it's clearly a mistake by the programmer, so a user should not see it.

File: lib/Target/ExportSystemC/Patterns/BuiltinEmissionPatterns.cpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    switch (bitWidth) {
+    case 1:
+      p << "bool";
+      break;
+    case 8:
+    case 16:
+    case 32:
+    case 64:
+      p << (type.isSigned() ? "" : "u") << "int" << bitWidth << "_t";
+      break;
```
[the remark this answers is not in the document]
Comment (maerhart) [reply lib/Target/ExportSystemC/Patterns/BuiltinEmissionPatterns.cpp:67]: The printer does not guarantee that. The idea is that the perpare-for-emission pass (that should be guaranteed to be executed before each invokation of the printer) checks if an integer type is present and makes sure that exactly one `#include <stdint.h>` is present.

File: lib/Target/ExportSystemC/Patterns/EmitCEmissionPatterns.cpp (hunk 1 above)
Comment (maerhart) [reply lib/Target/ExportSystemC/Patterns/EmitCEmissionPatterns.cpp:151]: Indeed, in many cases it's preferable to emit a `static_cast` to have the improved safety of additional compiler checks, but a C-style cast seemed 'more default' to me. Ideally, there would be an attribute on `emitc.cast` that specifies which of the several casts that C++ supports should be used.

File: test/Target/ExportSystemC/basic.mlir (hunk 2 above)
Comment (maerhart) [reply test/Target/ExportSystemC/basic.mlir:200]: I've updated the tests to be more modular and thus more maintainable. I've also added comments to state the reason why each test was added such that other contributors and my future self have an easier time deciding what updates/erasures to the tests are fine.

**Pull request merged**